README
Text-WrapProp version 0.04
==========================

Text::WrapProp - proportional line wrapping to form simple paragraphs

Text::WrapProp::wrap_prop() is a very simple paragraph formatter for proportional
text. It formats a single paragraph at a time by breaking lines at word boundaries.
You must supply the column width in floating point units which should be set to
the full width of your output device. A reference to a character width table
must also be supplied. The width units can be any metric you choose, as long as
the column width and the width table use the same metric.

Proportional wrapping is most commonly used in the publishing industry. In the
HTML age, custom proportional wrapping is less often performed as the browser
performs the calculations automatically.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

None

COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 1998, James Briggs

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Text::WrapProp.

0.01  Mon Sep 28 09:02:05 1998
	- original version; created by h2xs 1.18

0.02  Mon Sep 28 09:02:05 1998
	- minor cleanup for CPAN public use

0.03  Mon Sep 28 09:02:05 1998
	- minor cleanup for CPAN public use

0.04  Wed Oct 15 23:40:17 2014
	- modernize packaging to pass CPAN Testers



lib/Text/WrapProp.pm
package Text::WrapProp;

use strict;
use vars qw($VERSION @ISA @EXPORT @EXPORT_OK);

require Exporter;

@ISA = qw(Exporter AutoLoader);
# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

@EXPORT_OK = qw( wrap_prop );
@EXPORT = qw();

$VERSION = '0.04';

1;

sub wrap_prop {
   my ($text, $width, $ref_width_table) = @_;

   my @width_table = @{$ref_width_table};

   return '' if $text eq '';

   # simplify whitespace, including newlines
   $text =~ s/\s+/ /gs;

   my $cursor = 0; # width so far of line
   my $c;          # current character
   my $out;        # output buffer
   my $nextline = '';

   my $i=0;

   my $ltext = length $text;

# In 1998, regex scanning, but 50% slower than C-style char processing. eg. while ($text =~ /(.)/gcs) {
# In 2014, they're equally fast. See eg/NOTES.

   while ($i < $ltext) {
         # pop off next character
         $c    = substr($text,$i++,1);
         
         # don't need leading spaces at start of line
         next if $nextline eq '' and $c eq ' ';

         # see if character will fit on line - but don't include if too long
         if ($cursor + $width_table[ord $c] <= $width) {
            # another character fits
            $nextline .= $c;
            $cursor += $width_table[ord $c];
         }
         else {
            # find where we can wrap by checking backwards for separator
            my $j = length($nextline);
            foreach (split '', reverse $nextline) { # find separator
               $j--;
               last if /( |:|;|,|\.|-|\(|\)|\/)/o; # separator characters
            }

            # see if no separator found
            if (!$j) { # no separator, so just truncate line right here
               $i--; # rerun on $c
               $out .= $nextline."\n";
            }
            # 
            else { # separator found, so break line at separator
               $i -= length($nextline) - $j; # rerun characters after separator
               $out .= substr($nextline, 0, $j+1)."\n";
            }

            $nextline = '';
            $cursor = 0;
         }
   }

   $out.$nextline;
}

__END__

=head1 NAME

Text::WrapProp - proportional line wrapping to form simple paragraphs

=head1 SYNOPSIS 

	use Text::WrapProp qw(wrap_prop);

	print wrap_prop($text, $width, $ref_width_table);

=head1 DESCRIPTION

Text::WrapProp::wrap_prop() is a very simple paragraph formatter
for proportional text. It formats a
single paragraph at a time by breaking lines at word boundries.
You must supply the column width in floating point units which should
be set to the full width of your output device. A reference to a
character width table must also be supplied. The width units
can be any metric you choose, as long as the column width and
the width table use the same metric.

Proportional wrapping is most commonly used in the publishing
industry. In the HTML age, custom proportional wrapping is less often
performed as the browser performs the calculations automatically.

=head1 EXAMPLE

	use strict;
	use diagnostics;
	
	use Text::WrapProp qw(wrap_prop);

	my @width_table = 0.05 x 256;
	print wrap_prop("This is a bit of text that forms a normal book-style paragraph. Supercajafrajalisticexpialadocious!", 4.00, \@width_table);

=head1 BUGS

It's not clear what the correct behavior should be when WrapProp() is
presented with a word that is longer than a line.  The previous 
behavior was to die.  Now the word is split at line-length.

=head1 AUTHOR

James Briggs <james.briggs@yahoo.com>. Styled after Text::Wrap.

=cut


eg/wrap_prop-regex_slow.pl
#!/usr/bin/perl

# Program: wrap_prop-regex_slow.pl
# Author: James Briggs
# Date: Mon Sep 28 09:02:05 1998
# Env: Perl5
# Purpose: regex-powerered but slow version of the Text::WrapProp::wrap_prop subroutine

   use strict;
   use diagnostics;

   my @width_table = (0.05) x 256;

   print wrap_prop("This is a bit of text that forms a normal book-style paragraph. Supercajafrajalisticexpialadocious!", 4.00, \@width_table);

sub wrap_prop {
   my ($text, $width, $ref_width_table) = @_;

   my @width_table = @$ref_width_table;

   return '' if $text eq '';

   # simplify whitespace, including newlines
   $text =~ s/\s+/ /gs;

   my $cursor = 0; # width so far of line
   my $out;        # output buffer
   my $nextline = '';

   while ($text =~ /(.)/gcs) {
         
         # don't need leading spaces at start of line
         next if $nextline eq '' and $1 eq ' ';

         # see if character will fit on line - but don't include if too long
         if ($cursor + $width_table[ord $1] <= $width) {
            # another character fits
            $nextline .= $1;
            $cursor += $width_table[ord $1];
         }
         else {
            # find where we can wrap by checking backwards for separator
            my $j = length($nextline);
            foreach (split '', reverse $nextline) { # find separator
               $j--;
               last if /( |:|;|,|\.|-|\(|\)|\/)/o; # separator characters
            }

            # see if no separator found
            if (!$j) { # no separator, so just truncate line right here
               pos($text)--; # rerun on $1
               $out .= $nextline."\n";
            }
            # 
            else { # separator found, so break line at separator
               pos($text) -= length($nextline) - $j; # rerun characters after separator
               $out .= substr($nextline, 0, $j+1)."\n";
            }

            $nextline = '';
            $cursor = 0;
         }
   }

   $out.$nextline;
}

